### PR TITLE
Search: Fix indexing - re-index after initial provisioning

### PR DESCRIPTION
**Pull request by FZambia**
**What this PR does / why we need it**:

A safer alternative to #50517. Introducing entity events seems a rather huge and accurate refactoring, having signal like this should solve the biggest problem when dashboards not available in navigation after provisioning on fresh database.  But I am thinking more and more about triggers..

### CODE REVIEW

File: pkg/services/provisioning/dashboards/dashboard.go [hunk 1]
```diff
@@ -33,6 +34,10 @@ type Provisioner struct {
 	provisioner        dashboards.DashboardProvisioningService
 }
 
+func (provider *Provisioner) HasDashboardSources() bool {
+	return len(provider.fileReaders) > 0
```
Comment (ArturWierzbicki) [inline pkg/services/provisioning/dashboards/dashboard.go:38]: is it possible that we delete dashboards from the DB if their "corresponding file" on the disk is missing?

Comment (FZambia) [reply pkg/services/provisioning/dashboards/dashboard.go:38]: Yes, there is method called [CleanUpOrphanedDashboards](https://github.com/grafana/grafana/blob/main/pkg/services/provisioning/dashboards/dashboard.go#L80) that does this on every dashboard provision routine, but here we are not covering it – so will have some stale data in index for a while.

Comment (ArturWierzbicki) [reply pkg/services/provisioning/dashboards/dashboard.go:38]: okk, I was thinking about the `HasDashboardSources` name, specifically whether it means "database state will change once provisioning is complete" but I guess it's not so simple!

File: pkg/services/searchV2/index.go [hunk 2]
```diff
@@ -130,12 +130,27 @@ func (i *dashboardIndex) run(ctx context.Context, orgIDs []int64) error {
...
 		select {
 		case doneCh := <-i.syncCh:
+			// Executed on search read requests to make sure index is consistent.
 			lastEventID = i.applyIndexUpdates(ctx, lastEventID)
 			close(doneCh)
 		case <-partialUpdateTimer.C:
+			// Periodically apply updates collected in entity events table.
 			lastEventID = i.applyIndexUpdates(ctx, lastEventID)
 			partialUpdateTimer.Reset(partialUpdateInterval)
+		case <-reIndexSignalCh:
```
Comment (ArturWierzbicki) [inline pkg/services/searchV2/index.go:140]: isn't this branch functionally equivalent to `case <-fullReIndexTimer.C:`? as in, would this work
```
case <-reIndexSignalCh:
     log.info(...)
     fallthrough
case <-fullReIndexTimer.C:
     ....
```

**Review by ArturWierzbicki — APPROVED** [review]
didn't test, but the code looks good for a temporary bandaid :)

File: pkg/services/searchV2/index.go (hunk 2 above)
Comment (FZambia) [reply pkg/services/searchV2/index.go:140]: You are right - almost identical, ... Though `fallthrough` is not available in `select`.

Maybe refactor it like this:

```go
		case <-reIndexSignalCh:
			// External systems may trigger re-indexing, at this moment provisioning does this.
			i.logger.Info("Full re-indexing due to external signal")
			fullReIndexTimer.Reset(0)
```

File: pkg/services/provisioning/dashboards/dashboard.go (hunk 1 above)
Comment (FZambia) [reply pkg/services/provisioning/dashboards/dashboard.go:38]: Yep, require one more iteration in the future and only solving part of issues with provisioning. "database state will change once provisioning is complete" - I also thought about having this, but have not found an elegant way.

File: pkg/services/searchV2/index.go (hunk 2 above)
Comment (FZambia) [reply pkg/services/searchV2/index.go:143]: @ArturWierzbicki refactored a bit in ae8b70bc730802b19f143fd491e4bab1e177e1df

Comment (ArturWierzbicki) [reply pkg/services/searchV2/index.go:143]: looks good!

**Pull request merged**